### PR TITLE
Move Gridicon to @automattic/components

### PR DESCRIPTION
**Pull request by jsnajdr**
Moves the (very popular) `Gridicon` component to `@automattic/components`. The main motivation is that it's used by many components that we want to extract to separate packages: Happychat, Inline Help, ...

Best reviewed commit by commit, as the one that migrates the `import` statement is really big, and boring at the same time.

### CODE REVIEW

File: client/components/gridicon/index.tsx [hunk 1]
```diff
@@ -6,41 +6,38 @@ import {
...
+	} );
+
 	return (
 		<svg
 			xmlns="http://www.w3.org/2000/svg"
 			viewBox="0 0 24 24"
 			className={ iconClass }
 			height={ size }
 			width={ size }
-			onClick={ onClick }
```
Comment (sgomes) [inline client/components/gridicon/index.tsx:43]: I'm guessing this was removed simply because it's not being used anywhere?

**Review by sgomes — APPROVED** [review]
Thank you, @jsnajdr, this generally LGTM 👍 

My only concern with splitting this out into `@automattic/components` is what happens to the build process aspects. Specifically, this code currently relies on webpack being able to transform an import of `gridicons/svg-sprite/gridicons.svg` into a string with the absolute path of that SVG file, to be referenced at runtime. This is a very non-standard use of ES imports, so I'm not sure if it's a reasonable concern to push onto non-Calypso consumers.

Maybe the string should be passed into the component, which for the sake of simplicity could be done with a thin wrapper in `lib/`, keeping the import transformation logic exclusively inside Calypso? Or perhaps I'm being difficult and that's just something we're not too worried about at this point? 🙂

File: client/components/gridicon/index.tsx (hunk 1 above)
Comment (jsnajdr) [reply client/components/gridicon/index.tsx:43]: The `onClick` prop is passed as part of `...otherProps`. In other words, I changed
```js
const { icon, onClick, ...otherProps } = props;
return <svg onClick={ onClick } { ...otherProps } />;
```
to
```js
const { icon, ...otherProps } = props;
return <svg { ...otherProps } />;
```

**Comment by jsnajdr** [comment]
> this code currently relies on webpack being able to transform an import of gridicons/svg-sprite/gridicons.svg into a string with the absolute path of that SVG file, to be referenced at runtime.

Thanks @sgomes, that's a valid concern. I'll have to look into it. If bundlers can handle it, I guess it's OK to publish a package that exposes ES modules with asset imports. And document that the consumer is expected to cope with that.

File: client/components/gridicon/index.tsx (hunk 1 above)
Comment (sgomes) [reply client/components/gridicon/index.tsx:43]: Good point, I missed that 👍

**Comment by jsnajdr** [comment]
@sgomes I found there is already a `ProductIcon` component in the `@automattic/components` package, which also internally imports SVG assets (product icons) and exposes them as a React component that does `` internally. So I guess doing the same thing in `Gridicons` won't introduce any new issues.

@noahtallen The differences between this `Gridicon` component and the one from the `gridicons` package are:

The `@automattic/components` implementation is more efficient: it loads a SVG sprite with all the icons from a separate URL (can be cached) and shows one specific icon from that sprite. The actual JS/React code is very small.

On the other hand, the `gridicons` implementation contains a big `switch` statement with `case`s for each supported icons, and the SVG markup is transpiled into a tree of `React.createElement` calls, which expands the code size a lot.

The `@automattic/components` implementation doesn't support sub-components for individual icons, like `GridiconAddImage`.

The SVG sprite asset is 57kb and can be cached. The `gridicons` all-in-one React component is 102kB, and can't be cached.

The `@automattic/components` implementation also supports a `title` prop, providing some a11y.

The `@automattic/components` implementation could be incorporated into `gridicons`, and warrant a major 4.0.0 release. The biggest obstacle I guess is that we don't know who all the `gridicons` consumers are and what are their needs.

**Comment by sgomes** [comment]
> The SVG sprite asset is 57kb and can be cached. The `gridicons` all-in-one React component is 102kB, and can't be cached.

It's also important to note that even if they were the same size and had the same caching characteristics, it would still be vastly preferable to use the SVG external approach. JS is byte-for-byte the worst resource for a browser to handle, as it involves a lot of computation that has to be performed on slow, general-purpose CPU cores, while blocking the main thread. By contrast, image processing is often GPU or SOC-accelerated and happens in parallel to other tasks.

Furthermore, when the JS runs, the SVG still has to be parsed and rendered as if it had been loaded externally, so you're not even saving anything. The JS approach is always strictly worse, where performance is concerned.

**Comment by noahtallen** [comment]
> The biggest obstacle I guess is that we don't know who all the gridicons consumers are and what are their needs.

Makes sense :+1: Listed on npm, I see:

1. `@media-manager/components`
 2.   `@woocommerce/experimental`
 3.   `@automattic/components`
 4.   `@jonathansadowski/wpc-test`
 5.   `@woocommerce/components`
 6.   `@automattic/simple-components`
 7.  `@jonathansadowski/wpc-test`
 8.   `@woocommerce/block-library`
 9.  `happychat-client`
 10. `@woocommerce/onboarding`

And then probably some internal use cases. So it seems we'd mostly have to double check with WC to see if we could share the improvement everywhere

**Pull request merged**